### ISSUE TITLE
Training a UDE from CustomDerivatives doesn't use the supplied model

### ISSUE THREAD

**Issue by macquarrielucas** [Jack-H-Buckner/UniversalDiffEq.jl#155]
Hello,

I'm trying to train a UDE 
$$x' = rx - N(x,y,\theta) \\
y' = N(x,y,\theta)- my $$
on the data generated from 
$$x' = rx - xy/2 \\
y' = xy/2 - my $$

I made a `UDE` using`CustomDerivatives` inserting a neural network into my ODE, but my results were suspiciously good as i trained the model. I was getting the same results with any neural network I passed to `CustomDerivatives`, so I think `UDE.parameters.uhat` is training with a model that's different than the one I supplied. I've been looking through the docs and the source code, but I can't seem to figure out what's going on. The parameters of the model I supply also don't seem to be being updated. Is this the intended behavior?

Below is my code that can reproduce this. It will plot the predicted solution from `uhat` against the true solution, and also compare $N(x,y,\theta)$ with $xy/2$ 

```

using Lux, LuxCore
using Flux: mae, update!, mean
using Random
using Zygote 
using OrdinaryDiffEq
using UniversalDiffEq, DataFrames
using ProgressBars, Printf, Plots
using Optimization, Optimisers


## Plotting#########
#This will plot a 2x2 grid graphing 
#1. the true connecting function h(x,y) = 0.5xy, 
#2. the Neural Network, 
#3. phase plane including the ground truth solution and the estimatedsolution, 
#4. and time series of the ground truth solution and the estimated solution.


# Create grid for interaction function visualization
x = range(0, 3, length=20)
y = range(0, 3, length=20)
xy = [(i,j) for i in x, j in y]
# True interaction function
true_h = [0.5*i*j for (i,j) in xy]

function save_training_frame(model, p, iter, training_dir; save=false)
    # Create 2x2 grid plot
    plt = plot(layout=(2,2), size=(1600, 1200), titlefontsize=12)

    #prediction
    nn_h = [Neural_Network([i,j], pM, stM)[1][] for (i,j) in xy]

    # Get predictions for trajectories
    nn_sol = model.parameters.uhat
    

    # Top row: Interaction function surfaces
    plot!(plt[1], x, y, true_h, st=:surface, 
          title="True Interaction: h(x,y) = 0.5xy",
          xlabel="x", ylabel="y", zlabel="h(x,y)",
          zlim=(0,4.5), camera=(iter, 30))
    
    plot!(plt[2], x, y, nn_h, st=:surface, 
          title="Learned Interaction (Iteration $iter)",
          xlabel="x", ylabel="y", zlabel="h(x,y)",
          zlim=(0,4.5), camera=(iter, 30))
    
    # Bottom left: Phase plane
    plot!(plt[3], model.data[1,:], model.data[2,:], 
          label="True Solution", lw=2, color=:black,
          xlabel="x", ylabel="y",
          xlim=(0,4.5),ylim=(0,4.5),
           title="Phase Plane")
    plot!(plt[3], nn_sol[1,:], nn_sol[2,:], 
          label="Prediction", lw=2, color=:red, linestyle=:dash)
    scatter!(plt[3], [model.data[1,1]], [model.data[2,1]], color=:blue, label="Initial Condition")
    
    # Bottom right: Time series comparison
    plot!(plt[4], model.times, model.data[1,:], label="True x", color=:blue, lw=2)
    plot!(plt[4], model.times, model.data[2,:], label="True y", color=:green, lw=2)
    plot!(plt[4], model.times, nn_sol[1,:], label="KAN x", color=:blue, linestyle=:dash, lw=2)
    plot!(plt[4], model.times, nn_sol[2,:], label="KAN y", color=:green, linestyle=:dash, lw=2)
    vspan!(plt[4], [tspan_train[1], tspan_train[2]], alpha=0.2, color=:gray, label="Training Region")
    plot!(plt[4], 
        xlabel="Time", ylabel="Population",
        ylim=(0,6),
        title="Time Series Comparison", legend=:topright)
    if save
        # Save figure with iteration number
        savefig(plt, joinpath(training_dir, "frame_$(lpad(iter, 5, '0')).png"))
    end
    return plt
end

##Ground Truth Data Generation ########
function lotka!(du,u,p,t) 
    du[1] = p.r*u[1] - 0.5f0*u[1]*u[2]
    du[2] = 0.5f0*u[1]*u[2] - p.m*u[2]
end

#data generation parameters
dt=0.1
tspan = (0.0, 14)
tspan_train=(0.0, 14)
u0 = [1.0f0, 1.0f0]
p_= (r= 1.0f0, m=1.0f0)
prob = ODEProblem(lotka!, u0, tspan, p_)

solution = solve(prob, Tsit5(), abstol = 1e-12, reltol = 1e-12, saveat = dt, verbose = false )
# Extract solution data
times = solution.t  # Time points
u1 = solution[1, :]  # First variable (e.g., prey population)
u2 = solution[2, :]  # Second variable (e.g., predator population)
# Create a DataFrame
data = DataFrame(time=times, prey=u1, predator=u2)

########Neural Network Model definition#####
dims_in = 2; dims_out = 1; hidden = 1
Neural_Network = Lux.Chain(Lux.Dense(dims_in,hidden), Lux.Dense(hidden,dims_out))
# initialize the neural network states and parameters
rng = Random.default_rng()
pM, stM = Lux.setup(rng,Neural_Network)
# Define the UDE
function lotka_voltera!(du,u,p,t)
    C, _ = Neural_Network(u,pM,stM)
    du[1] = p.r*u[1] - C[1]
    du[2] = C[1] - p.m*u[2]
end

initial_parameters = (NN=pM, r=2.0, m=0.5)
model = CustomDerivatives(data, lotka_voltera!, initial_parameters)

##Training
N_iter = 100
iterator = ProgressBar(1:N_iter)
iters_per_loop = 10
@time for i in 1:N_iter#iterator

    #Update model
    # set optimization problem 
    target  = (x,p) -> model.loss_function(x)
    adtype  = Optimization.AutoZygote()
    optf    = Optimization.OptimizationFunction(target, adtype)
    optprob = Optimization.OptimizationProblem(optf, model.parameters)

    # run optimizer for 10 steps
    @time sol = Optimization.solve(optprob, 
                                Optimisers.Adam(0.05), 
                                maxiters = iters_per_loop)
    
    # assign parameters to model 
    model.parameters = sol.u

    # CALLBACK
    loss_curr=deepcopy(model.loss_function(model.parameters))
    #loss_curr_test=deepcopy(loss_test(p))
    #@time set_description(iterator, string("Iter:", i, "| Loss:", @sprintf("%.2f", loss_curr), "|"))

    display(save_training_frame(model, pM, i, "frames_2"; save=false))
    print("Hello?!?!?")
    print(pM)
    #plt = UniversalDiffEq.plot_state_estimates(model)
    # SAVE
    #callback(i)
end
```

**Comment by Jack-H-Buckner** [comment]
Hi, and thanks for testing out the package! 

My guess is that the package is working as intended, and your results would make more sense if you used additional functions to evaluate the model's performance. 

Check out the `plot_predictions` function. The base training routine in the package does two things. It estimates a set of state variables (the uhats you found in the source code), assuming the data include some observaiton error. It also trains the parameters of the UDE model to predict the changes observed in the estimated states (uhat).

The `plot_state_estimates` function in your code shows the correspondence between the estimated states and the data, which is usually pretty good regardless of the model. The `plot_predictions` function will evaluate the model's performance, predicting one step ahead within the sample. A poorly specified model will usually not make very good predictions and you will be able to catch that using the `plot_predicitons` function. 

A good way to test this is by passing dudt = 0 to CustomDerivatives. if the package works the way we expect, the `plot_state_estimates` will show reasonably close agreement between the estimated states and the data, but  `plot_predictions` will show a line of points that do not correspond with the changes in the data set. 

Please let us know if this fixes the issues you are having. 

Best,
Jack

**Comment by macquarrielucas** [comment]
Passing dudt=0 to CustomeDerivatives does as you predicted.

As I see, the uhats are set in Optimizers.jl by derivative_matching_loss in LossFunctionConstructors.jl, but are then interpolated by interpolate_derivs. Where exactly are these uhats being computed? 

Please correct me if I'm wrong, but are the uhats not supposed to be the simulated solution of the UDE? The simulated solutions is really the only data that I'm looking for.

Maybe I am misunderstanding the documentation. Do you have a reference for these observation model and process models? 

Thanks for helping me and making the package! I was making my own script before discovering thing, but you include a lot of features I don't have to figure out for myself :)

**Comment by Jack-H-Buckner** [comment]
Yes, check out the pre-print that should answer a lot of questions. Also, please note that the current version of the package on the Julia registry only includes one training routine (as described in the pre-print). However, we have recently added several other options that can be accessed through the `train!` function [link to docs](https://jack-h-buckner.github.io/UniversalDiffEq.jl/dev/API/#UniversalDiffEq.train!-Tuple%7BUDE%7D). Some of the options follow tutorials from the DiffEqFlux docs pretty closely and might behave a bit more like what you are expecting. 

to use `train!` you will need to load the development version of the package `add https://github.com/Jack-H-Buckner/UniversalDiffEq.jl.git`.

**Comment by macquarrielucas** [comment]
Ok I've familiarized myself more with the source code and the theory.  I see that the `uhat`s that are used in the `plot_state_estimates` function  are calculated based on which loss_function is chosen, but it seems like `CustomDerivatives` is using it's own loss function based on `init_loss` in `helpers.jl`, where the dynamic loss is 
```
        # dynamics loss 
        L_proc = 0
        for t in 2:(size(data)[2])
            u0 = parameters.uhat[:,t-1]
            u1 = parameters.uhat[:,t]
            dt = times[t]-times[t-1]
            u1hat, epsilon = process_model.predict(u0,times[t-1],dt,parameters.process_model) 
            L_proc += process_loss.loss(u1,u1hat,dt,parameters.process_loss)
        end
```
is there any reason that this loss is defined differently than the ones in `LossFunctionConstructors.jl`? If I understand right, this should be multiple shooting with a single prediction step. 

I think what confused me is that I didn't realize that the state estimate predictions were being calculated based on which loss function was selected with the model. In my mind, a process model should be independent of which loss function is used to train the neural network / parameters that are inside of it.  Is there a reason that I'm missing for putting them together?

Anyhow, I should have been using the forecast function to simulate a solution from the process model beginning at the first data points.

**Comment by Jack-H-Buckner** [comment]
I'm sorry for not getting back to you sooner. The init_loss function in helpers has been superseded by the loss function in loss function constructors, but we kept the old version to avoid making a breaking change.
